crates/c-api/include/wasmtime.h

```diff
@@ -55,53 +67,46 @@ WASMTIME_CONFIG_PROP(bool, profiler, wasmtime_profiling_strategy_t)
 //   it. This will be parsed and converted to the binary format.
 // * `ret` - if the conversion is successful, this byte vector is filled in with
 //   the WebAssembly binary format.
-// * `error_message` - if the conversion fails, this is filled in with a
-//   descriptive error message of why parsing failed. This parameter is
-//   optional.
 //
-// Returns `true` if conversion succeeded, or `false` if it failed.
-WASM_API_EXTERN bool wasmtime_wat2wasm(
+// Returns a non-null error if parsing fails, or returns `NULL`. If parsing
+// fails then `ret` isn't touched.
+WASM_API_EXTERN own wasmtime_error_t* wasmtime_wat2wasm(
     const wasm_byte_vec_t *wat,
-    own wasm_byte_vec_t *ret,
-    own wasm_byte_vec_t *error_message
+    own wasm_byte_vec_t *ret
 );
 
 ///////////////////////////////////////////////////////////////////////////////
 //
 // wasmtime_linker_t extension type, binding the `Linker` type in the Rust API
 
-#define WASMTIME_DECLARE_OWN(name) \
-  typedef struct wasmtime_##name##_t wasmtime_##name##_t; \
-  \
-  WASM_API_EXTERN void wasmtime_##name##_delete(own wasmtime_##name##_t*);
-
 WASMTIME_DECLARE_OWN(linker)
 
 WASM_API_EXTERN own wasmtime_linker_t* wasmtime_linker_new(wasm_store_t* store);
 
 WASM_API_EXTERN void wasmtime_linker_allow_shadowing(wasmtime_linker_t* linker, bool allow_shadowing);
 
-WASM_API_EXTERN bool wasmtime_linker_define(
+WASM_API_EXTERN own wasmtime_error_t* wasmtime_linker_define(
```

Comment:
This (and the similar changes above) is the source of the .NET CI failures since it's now the inverse of what the .NET code is expecting (i.e. return value of 0 is now success rather than failure).

As I'm now moving the .NET code to `wasmtime-dotnet`, so we just merge this in with failing CI and fix it there or do you want to get this CI green and I'll port the fix to the other repo?

Reply:
Indeed! Mind double-checking my C# to confirm that what's there is reasonable?

(it's somewhat intentionally not fully fleshed out in every spot since I figured you'll want to tweak APIs further as well)